### PR TITLE
Ignore timeout error while polling operation status from ARM

### PR DESCRIPTION
**Pull request by dikhakha**
Ignore timeout error when we poll for ARM operation status. This is to avoid failing in case poll operation request timed out due to ARM taking long time to process the request. The operation could have ended up successfully however the tasks fails because of this.

### CODE REVIEW

**Comment by dikhakha** [comment]
@bishal-pdMSFT / @arjgupta / @vinodkumar3  Can you please have a look at this PR ?

File: Tasks/Common/azure-arm-rest-v2/package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "azure-arm-rest-v2",
-  "version": "2.0.0",
+  "version": "2.0.1",
```
Comment (bishal-pdMSFT) [inline Tasks/Common/azure-arm-rest-v2/package.json]: I don;t think we do update the common package version

File: Tasks/Common/azure-arm-rest-v2/AzureServiceClientBase.ts [hunk 2]
```diff
@@ -140,32 +140,46 @@ export class AzureServiceClientBase {
...
+                } else {
+                    break;
                 }
-                // If timeout; throw;
-                if (!waitIndefinitely && timeout < new Date().getTime()) {
-                    throw new Error(tl.loc("TimeoutWhileWaiting"));
+            }
+            catch (error) {
+                let errorString: string = (error && error.toString()) || "";
+                if(errorString && errorString.toLowerCase().indexOf("request timeout") >= 0 && ignoreTimeoutErrorThreshold > 0) {
```
Comment (bishal-pdMSFT) [inline Tasks/Common/azure-arm-rest-v2/AzureServiceClientBase.ts]: is there something more to look for than just "request timeout"?

File: Tasks/Common/azure-arm-rest-v2/AzureServiceClientBase.ts [hunk 3]
```diff
@@ -140,32 +140,46 @@ export class AzureServiceClientBase {
...
+                    tl.debug(`Request Timeout: ${request.uri}`);
+                    ignoreTimeoutErrorThreshold--;
                 }
-
-                // Retry after given interval.
-                var sleepDuration = 15;
-                if (response.headers["retry-after"]) {
-                    sleepDuration = parseInt(response.headers["retry-after"]);
+                else {
+                    throw error;
```
Comment (bishal-pdMSFT) [inline Tasks/Common/azure-arm-rest-v2/AzureServiceClientBase.ts]: just `throw`. Not sure about javascript but in C# `throw error` would change the call stack

File: Tasks/Common/azure-arm-rest-v2/AzureServiceClientBase.ts [hunk 4]
```diff
@@ -140,32 +140,46 @@ export class AzureServiceClientBase {
...
+                    await this.sleepFor(sleepDuration);
+                } else {
+                    break;
                 }
-                // If timeout; throw;
-                if (!waitIndefinitely && timeout < new Date().getTime()) {
-                    throw new Error(tl.loc("TimeoutWhileWaiting"));
+            }
+            catch (error) {
+                let errorString: string = (error && error.toString()) || "";
```
Comment (vinodkumar3) [inline Tasks/Common/azure-arm-rest-v2/AzureServiceClientBase.ts]: (!!error && error.toString())

File: Tasks/Common/azure-arm-rest-v2/AzureServiceClientBase.ts (hunk 2 above)
Comment (vinodkumar3) [inline Tasks/Common/azure-arm-rest-v2/AzureServiceClientBase.ts]: !!errorString

File: Tasks/Common/azure-arm-rest-v2/package.json (hunk 1 above)
Comment (vinodkumar3) [reply Tasks/Common/azure-arm-rest-v2/package.json]: +1

File: Tasks/AzureResourceGroupDeploymentV2/package.json [hunk 5]
```diff
@@ -2,7 +2,7 @@
   "name": "AzureResourceGroupDeployment",
   "main": "main.js",
   "dependencies": {
-    "azure-arm-rest-v2": "file:../../_build/Tasks/Common/azure-arm-rest-v2-2.0.0.tgz",
+    "azure-arm-rest-v2": "file:../../_build/Tasks/Common/azure-arm-rest-v2-2.0.1.tgz",
```
Comment (vinodkumar3) [inline Tasks/AzureResourceGroupDeploymentV2/package.json]: No need to update the version

**Comment by vinodkumar3** [comment]
I think you might need to update the patch version of many tasks depending upon task common lib

File: Tasks/Common/azure-arm-rest-v2/AzureServiceClientBase.ts (hunk 3 above)
Comment (dikhakha) [reply Tasks/Common/azure-arm-rest-v2/AzureServiceClientBase.ts]: In javascript, you can re throw the error which will preserve the stack by doing: `throw error`. `throw new Error(error)` will lead to stack trace being lost.

File: Tasks/Common/azure-arm-rest-v2/AzureServiceClientBase.ts (hunk 2 above)
Comment (dikhakha) [reply Tasks/Common/azure-arm-rest-v2/AzureServiceClientBase.ts]: The error is of form "Request Timeout: <url>", so I think this should be good enough.

**Comment by dikhakha** [comment]
> I think you might need to update the patch version of many tasks depending upon task common lib

This functionality is not absolutely required by other tasks who are using this module. Do you think there is need to update other tasks version ?

**Comment by dikhakha** [comment]
> > I think you might need to update the patch version of many tasks depending upon task common lib
> 
> This functionality is not absolutely required by other tasks who are using this module. Do you think there is need to update other tasks version ? I will update for this task

I see that build itself will fail if tasks consuming this common is not updated. Will update other task versions as well. Thanks

**Pull request merged**